include/libopencm3/stm32/fdcan.h

```diff
@@ -0,0 +1,855 @@
+
+/* REL[3:0]: FDCAN core release number */
+#define FDCAN_CREL_REL_SHIFT			28
+#define FDCAN_CREL_REL_MASK				(__BITS(4) << FDCAN_CREL_REL_SHIFT)
+
+
+/* DSJW[3:0]: Synchronization jump width */
+#define FDCAN_DBTP_DSJW_SHIFT			0
+#define FDCAN_DBTP_DSJW_MASK			(__BITS(4) << FDCAN_DBTP_DSJW_SHIFT)
+
+#define FDCAN_DBTP_DSJW_TQ(tq)			((tq) << FDCAN_DBTP_DSJW_SHIFT)
+
+/* DTSEG2[3:0]: Data time segment after sample point */
+#define FDCAN_DBTP_DTSEG2_SHIFT			4
+#define FDCAN_DBTP_DTSEG2_MASK			(__BITS(4) << FDCAN_DBTP_DTSEG2_SHIFT)
+
+#define FDCAN_DBTP_DTSEG2_TQ(tq)		((tq) << FDCAN_DBTP_DTSEG2_SHIFT)
+
+/* DTSEG1[4:0]: Data time segment before sample point */
+#define FDCAN_DBTP_DTSEG1_SHIFT			8
+#define FDCAN_DBTP_DTSEG1_MASK			(__BITS(5) << FDCAN_DBTP_DTSEG1_SHIFT)
+
+#define FDCAN_DBTP_DTSEG1_TQ(tq)		((tq) << FDCAN_DBTP_DTSEG1_SHIFT)
+
+/* DBRP[4:0]: Data bit rate prescaler */
+#define FDCAN_DBTP_DBRP_SHIFT			16
+#define FDCAN_DBTP_DBRP_MASK			(__BITS(5) << FDCAN_DBTP_DBRP_SHIFT)
+
+#define FDCAN_DBTP_TDC					(1 << 23)
+
+#define FDCAN_TEST_LBCK					(1 << 4)
+/* TX[1:0]: Control of transmit pin */
+#define FDCAN_TEST_TX_SHIFT				5
+#define FDCAN_TEST_TX_MASK				(__BITS(2) << FDCAN_TEST_TX_SHIFT)
+
+#define FDCAN_TEST_RX					(1 << 7)
+
+/* WDC[7:0]: RAM watchdog configuration */
+#define FDCAN_RWD_WDC_SHIFT				0
+#define FDCAN_RWD_WDC_MASK				(__BITS(8) << FDCAN_RWD_WDC_SHIFT)
+
+/* WDV[7:0]: RAM watchdog actual value */
+#define FDCAN_RWD_WDV_SHIFT				7
+#define FDCAN_RWD_WDV_MASK				(__BITS(8) << FDCAN_RWD_WDV_SHIFT)
+
+/** @defgroup fdcan_cccr FDCAN CC control register bits
+ * @{
+ */
+#define FDCAN_CCCR_INIT					(1 << 0)
+#define FDCAN_CCCR_CCE					(1 << 1)
+#define FDCAN_CCCR_ASM					(1 << 2)
+#define FDCAN_CCCR_CSA					(1 << 3)
+#define FDCAN_CCCR_CSR					(1 << 4)
+#define FDCAN_CCCR_MON					(1 << 5)
+#define FDCAN_CCCR_DAR					(1 << 6)
+#define FDCAN_CCCR_TEST					(1 << 7)
+#define FDCAN_CCCR_FDOE					(1 << 8)
+#define FDCAN_CCCR_BRSE					(1 << 9)
+#define FDCAN_CCCR_PXHD					(1 << 12)
+#define FDCAN_CCCR_EFBI					(1 << 13)
+#define FDCAN_CCCR_TXP					(1 << 14)
+#define FDCAN_CCCR_NISO					(1 << 15)
+/**@}*/
+
+/* NTSEG2[6:0]: Nominal timing segment after sample point length */
+#define FDCAN_NBTP_NTSEG2_SHIFT			0
+#define FDCAN_NBTP_NTSEG2_MASK			(__BITS(7) << FDCAN_NBTP_NTSEG2_SHIFT)
+
+/** Obtain value for use as TSEG2 for normal speed.
+ * @param ts amount of tq
+ * @returns value suitable for use with fdcan_init
+ */
+#define FDCAN_NBTP_NTSEG2_TQ(tq)		((tq) << FDCAN_NBTP_NTSEG2_SHIFT)
+
+/* NTSEG1[7:0]: Nominal timing segment before sample point length */
+#define FDCAN_NBTP_NTSEG1_SHIFT			8
+#define FDCAN_NBTP_NTSEG1_MASK			(__BITS(8) << FDCAN_NBTP_NTSEG1_SHIFT)
+
+/** Obtain value for use as TSEG1 for normal speed.
+ * @param ts amount of tq
+ * @returns value suitable for use with fdcan_init
+ */
+#define FDCAN_NBTP_NTSEG1_TQ(tq)		((tq) << FDCAN_NBTP_NTSEG1_SHIFT)
```

Comment:
why would you not just let the user pass "tq" to fdcan_init directly and do the shift internally?

Reply:
I wanted to make API semantically as close to can.c as possible. can.c defines a lot of NTSEG_xTQ defines. But FDCAN is much more flexible with some prescalers as wide as 7 bits and having 256 defines would be weird. If API similarity is not desired, I may change it to accept bare numbers. Both here and on other places around the code, which behave similarly.

Reply:
I'm not entirely sure that following the existing can.c api is the best idea.  it's an extremely old piece of code, and very out of style with almost everything else in the library.  we receive almost no feedback on it, and it receies almost no changes.  Might be perfect, might be unusable garbage :)

Reply:
Removed, all uses converted to plain hexa values.